primitives/src/merkle_tree.rs

```diff
@@ -0,0 +1,99 @@
+use std::fmt;
+use std::hash::Hasher;
+use std::iter::FromIterator;
+use crypto::sha3::{Sha3, Sha3Mode};
+use crypto::digest::Digest;
+use merkletree::hash::{Algorithm, Hashable};
+use merkletree::merkle;
+use merkletree::merkle::{VecStore};
+use merkletree::proof::Proof;
+use tiny_keccak::Keccak;
+use hex::ToHex;
+
+#[derive(Clone)]
+struct KeccakAlgorithm(Keccak);
+
+impl KeccakAlgorithm {
+    pub fn new() -> KeccakAlgorithm {
+        KeccakAlgorithm(Keccak::new_sha3_256())
+    }
+}
+
+impl Default for KeccakAlgorithm {
+    fn default() -> KeccakAlgorithm {
+        KeccakAlgorithm::new()
+    }
+}
+
+impl Hasher for KeccakAlgorithm {
+    #[inline]
+    fn write(&mut self, msg: &[u8]) {
+        self.0.update(msg)
+    }
+
+    #[inline]
+    fn finish(&self) -> u64 {
+        unimplemented!()
+    }
+}
+
+pub type MerkleItem = [u8; 32];
+
+impl Algorithm<MerkleItem> for KeccakAlgorithm {
+    #[inline]
+    fn hash(&mut self) -> MerkleItem {
+        let mut res: [u8; 32] = [0; 32];
+        self.0.clone().finalize(&mut res);
+        res
+    }
+
+    #[inline]
+    fn reset(&mut self) {
+        self.0 = Keccak::new_sha3_256();
+    }
+}
+
+pub struct MerkleTree(merkle::MerkleTree<MerkleItem, KeccakAlgorithm, VecStore<MerkleItem>>);
+
+impl MerkleTree {
+    pub fn new(leaves: Vec<MerkleItem>) -> MerkleTree {
+        let t = merkle::MerkleTree::from_iter(leaves);
```

Comment:
If you followed on my comment in the issue:
https://github.com/AdExNetwork/adex-validator-stack-rust/issues/84#issuecomment-510789831
The main issue I've bumped in to was that is asserts that there are more than 1 leaves, which means that the heartbeat MerkleTree we need to create won't be possible:

https://docs.rs/merkletree/0.10.0/src/merkletree/merkle.rs.html#915

https://github.com/AdExNetwork/adex-validator/blob/a3b59a4c43d02ab4fc9de8ed5c48957592a260d4/services/validatorWorker/heartbeat.js#L13

Reply:
Actually the entire implementation is wrong, including the previous one. The merkle tree implementation doesn't support 8byte arrays of 64 length which is the size of nodes/leafs in the JS implementation.